Improve RGB driver in pull #6808; solves #6968

santaimpersonator opened the pull request:
*By completing this PR sufficiently, you help us to review this Pull Request quicker and also help improve the quality of Release Notes*

### Checklist
1. [ ] Please provide specific title of the PR describing the change, including the component name (*eg. „Update of Documentation link on Readme.md“*)
2. [ ] Please provide related links (*eg. Issue which will be closed by this Pull Request*)
3. [ ] Please **update relevant Documentation** if applicable
4. [ ] Please check [Contributing guide](https://docs.espressif.com/projects/arduino-esp32/en/latest/contributing.html)

*This entire section above can be deleted if all items are checked.*

-----------

## Description of Change
This solution adds a `RGB_BUILTIN` variable to pull request https://github.com/espressif/arduino-esp32/pull/6808. With the original solution (https://github.com/espressif/arduino-esp32/pull/6808), users were limited to controlling a single built-in LED. By creating a new variable `RGB_BUILTIN`, users are able to control a regular LED and RGB LED with the `digitalWrite()` function, separately from each other.

* Replaces the use of the `LED_BUILTIN` variable for the RGB LED
* Removes the need for `BOARD_HAS_NEOPIXEL`
* Also, renames `LED_BRIGHTNESS` to match new variable name `RGB_BRIGHTNESS`

For boards that have a regular LED and RGB LED on their board, users can define the pin configuration *(below)* and the `Blink.ino` and `BlinkRGB.ino` example sketches would operate on the regular and RGB LEDs separately.

```
static const uint8_t LED_BUILTIN = 13;
static const uint8_t RGB_BUILTIN = SOC_GPIO_PIN_COUNT+2;
#define BUILTIN_LED  LED_BUILTIN // backward compatibility
#define LED_BUILTIN LED_BUILTIN
#define RGB_BUILTIN RGB_BUILTIN
#define RGB_BRIGHTNESS 65
```


## Tests scenarios
I have tested the proposed changes with the regular [`Blink.ino`](https://github.com/arduino/Arduino/blob/master/app/testdata/sketches/Blink/Blink.ino) and the **ESP32**>**GPIO**>[`BlinkRGB.ino`](https://github.com/espressif/arduino-esp32/blob/master/libraries/ESP32/examples/GPIO/BlinkRGB/BlinkRGB.ino) example sketches. These tests were performed on an ESP32-WROOM board with a regular LED and RGB LED. The sketches were also tested for the following pin configurations:

* Regular and RGB LED:
    * `Blink.ino` only controls regular LED
    * `BlinkRGB.ino` only controls the RGB LED
    ```
    static const uint8_t LED_BUILTIN = 13;
    static const uint8_t RGB_BUILTIN = SOC_GPIO_PIN_COUNT+2;
    #define BUILTIN_LED  LED_BUILTIN // backward compatibility
    #define LED_BUILTIN LED_BUILTIN
    #define RGB_BUILTIN RGB_BUILTIN
    #define RGB_BRIGHTNESS 65
    ```
* Only RGB LED:
    * `Blink.ino` and `BlinkRGB.ino` both control the RGB LED
    ```
    static const uint8_t LED_BUILTIN = SOC_GPIO_PIN_COUNT+2;
    #define BUILTIN_LED  LED_BUILTIN // backward compatibility
    #define LED_BUILTIN LED_BUILTIN
    #define RGB_BUILTIN LED_BUILTIN
    #define RGB_BRIGHTNESS 65
    ```
* Only regular LED:
    * Only `Blink.ino` works with the regular LED
    ```
    static const uint8_t LED_BUILTIN = 13;
    #define BUILTIN_LED  LED_BUILTIN // backward compatibility
    #define LED_BUILTIN LED_BUILTIN
    ```


## Related links
Closes https://github.com/espressif/arduino-esp32/issues/6968
*\*Based on original pull request (https://github.com/espressif/arduino-esp32/pull/6808) and issue (https://github.com/espressif/arduino-esp32/issues/6783)*

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6979) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6979) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/arduino-esp32?pullRequest=6979) it.</sub>

santaimpersonator commented:
Does the [GPIO documentation](https://docs.espressif.com/projects/arduino-esp32/en/latest/api/gpio.html?highlight=led#gpio-input-and-output-modes) need to be updated for the RGB driver?

SuGlider commented:
I have mixed feeling about this PR... It looks like it just changes the name of the symbols.
What exactly is the improvement @santaimpersonator?

PilnyTomas commented:
It is related to #6968 where it stated that some boards with both normal LED and RGB LED can use only the RGB. This PR will enable use both `LED_BUILTIN` and `RGB_BUILTIN`

santaimpersonator commented:
> I have mixed feelings about this PR... It looks like it just changes the name of the symbols. What exactly is the improvement @santaimpersonator?
> 
> > This solution adds a RGB_BUILTIN variable to pull request #6808.
> 
> It actually replaces, it doesn't add anything.

Yes, from the initial commit, it may seem like I have only replaced variable names. However, my pull request allows boards with a both a regular LED and RGB LED to be controlled with the `digitalWrite()` function. 

* The initial RGB driver only provided control on one of those LEDs when used with the `Blink.ino` and `BlinkRGB.ino` example codes.
* As shown in the test scenarios (above) that I provided, users can now control both LEDs separately *(if they wish to define the board that way)*
  > * Regular and RGB LED:
  >   
  >   * `Blink.ino` only controls regular LED
  >   * `BlinkRGB.ino` only controls the RGB LED
  >   
  >   ```
  >   static const uint8_t LED_BUILTIN = 13;
  >   static const uint8_t RGB_BUILTIN = SOC_GPIO_PIN_COUNT+2;
  >   #define BUILTIN_LED  LED_BUILTIN // backward compatibility
  >   #define LED_BUILTIN LED_BUILTIN
  >   #define RGB_BUILTIN RGB_BUILTIN
  >   #define RGB_BRIGHTNESS 65
  >   ```

I caught this issue with our board that we were about to submit a pull request for. Our board has both types of LEDs and the regular LED is already silked with `STAT`. With the current release, we can only define the board to work with one of the LEDs when using the `Blink.ino` and `BlinkRGB.ino` example codes.

* If we define the board with the regular LED (`STAT`), only the `Blink.ino` example is compatible with the board.
  ```
  static const uint8_t LED_BUILTIN = 13;
  #define BUILTIN_LED  LED_BUILTIN // backward compatibility
  #define LED_BUILTIN LED_BUILTIN
  ```
* If we define the board with `BOARD_HAS_NEOPIXEL`, both the `Blink.ino` and `BlinkRGB.ino` example codes are compatible with the board. However, the `Blink.ino` example blinks the RGB LED and not the the regular LED that is labeled with `STAT` (Status LED) on our board.
  ```
  static const uint8_t LED_BUILTIN = SOC_GPIO_PIN_COUNT+2;
  #define BUILTIN_LED  LED_BUILTIN // backward compatibility
  #define LED_BUILTIN LED_BUILTIN
  #define BOARD_HAS_NEOPIXEL
  #define LED_BRIGHTNESS 65
  ```

My changes basically have users define the current `BOARD_HAS_NEOPIXEL` variable with a value (1). I then renamed that variable to `RGB_BUILTIN` to align with the `LED_BUILTIN` variable name (2). Then, I renamed the `LED_BRIGHTNESS` variable to `RGB_BRIGHTNESS` match the `RGB_BUILTIN` name (3). To breakdown these changes, please refer to lines 137-141 in [cores/esp32/esp32-hal-gpio.c](https://github.com/espressif/arduino-esp32/pull/6979/files#diff-cc07ee5f3d34f8f2648dff856f751f53c9b84477993e8e6adb0f8291d897cc09)

* Original code:
    ```
    extern void ARDUINO_ISR_ATTR __digitalWrite(uint8_t pin, uint8_t val)
    {
        #ifdef BOARD_HAS_NEOPIXEL
            if(pin == LED_BUILTIN){
                //use RMT to set all channels on/off
                const uint8_t comm_val = val != 0 ? LED_BRIGHTNESS : 0;
                neopixelWrite(LED_BUILTIN, comm_val, comm_val, comm_val);
                return;
            }
        #endif
    ```
1. Have users define the `BOARD_HAS_NEOPIXEL` variable with a value. Then, use it to replace the value for `LED_BUILTIN`.
    ```
    extern void ARDUINO_ISR_ATTR __digitalWrite(uint8_t pin, uint8_t val)
    {
        #ifdef BOARD_HAS_NEOPIXEL
            if(pin == BOARD_HAS_NEOPIXEL){
                //use RMT to set all channels on/off
                const uint8_t comm_val = val != 0 ? LED_BRIGHTNESS : 0;
                neopixelWrite(BOARD_HAS_NEOPIXEL, comm_val, comm_val, comm_val);
                return;
            }
        #endif
    ```
2. Renamed the `BOARD_HAS_NEOPIXEL` variable to `RGB_BUILTIN` to align with use of the `LED_BUILTIN` in the example codes.
    ```
    extern void ARDUINO_ISR_ATTR __digitalWrite(uint8_t pin, uint8_t val)
    {
        #ifdef RGB_BUILTIN
            if(pin == RGB_BUILTIN){
                //use RMT to set all channels on/off
                const uint8_t comm_val = val != 0 ? LED_BRIGHTNESS : 0;
                neopixelWrite(RGB_BUILTIN, comm_val, comm_val, comm_val);
                return;
            }
        #endif
    ```
3. Renamed the `LED_BRIGHTNESS` variable to `RGB_BRIGHTNESS` match the `RGB_BUILTIN` name.
    ```
    extern void ARDUINO_ISR_ATTR __digitalWrite(uint8_t pin, uint8_t val)
    {
        #ifdef RGB_BUILTIN
            if(pin == RGB_BUILTIN){
                //use RMT to set all channels on/off
                const uint8_t comm_val = val != 0 ? RGB_BRIGHTNESS : 0;
                neopixelWrite(RGB_BUILTIN, comm_val, comm_val, comm_val);
                return;
            }
        #endif
    ```

**As you can see, the primary change is in (1), where the `BOARD_HAS_NEOPIXEL` variable is defined with a value. That is then used to replace the value for `LED_BUILTIN`.** The rest of the changes rename variables to align better with their use in the example codes.

SuGlider commented:
Thanks @santaimpersonator @PilnyTomas for the clarification. 
Indeed, it adds value.

it just makes me wonder that maybe we may need to review all the variants and their `arduino_pins.h` to add the right `#define xxx` considering each board hardware.... We shall document it ASAP.

santaimpersonator commented:
@SuGlider Awesome! Thanks for the approval.

**NOTE:** For reviewing the variant files:
* All the SparkFun variants should be fine as they are (with regards to this pull request). The pull request we are about to submit, will be our first ESP32 board with an RGB LED.
* In this pull request, I already changed the [`pins_arduino.h` files](https://github.com/espressif/arduino-esp32/pull/6979/commits/456a21adb99fea92c8e49d3626f048f89879e935) for the [three board variants in #6808](https://github.com/espressif/arduino-esp32/pull/6808/files)

santaimpersonator commented:
My coworker just pointed this out to me in the Arduino IDE:
![image](https://user-images.githubusercontent.com/36016723/179246151-014a5217-ff4c-4ca4-9ac6-672bd84f6f30.png)

Does this Arduino core have a `keywords.txt` file, where we can include `RGB_BUILTIN` as a keyword?
